File: ruby/missing_number.rb
def missing_number1(arr1, arr2)
  arr1-arr2
end

def missing_number2(arr1,arr2)
  arr2 = arr2.sort
  counter = 0
  while (arr1[counter] == arr2[counter])
    counter +=1
  end
  puts arr1[counter]
end


missing_number2([1,2,3,4,5,6,7],[3,7,2,1,4,6])

File: python/missing_number.py
# finder1 does not work for duplicate numbers

def finder1(arr1,arr2):
    arr1.sort()
    arr2.sort()

    # zip itereates over two lists in parallel:
    for x, y in zip(arr1,arr2):
        if x != y:
            return x


import collections

# finder2 does work with duplicate numbers.
def finder2(arr1, arr2):

    # Using default dict to avoid key errors
    d=collections.defaultdict(int)

    # Add a count for every instance in Array 1
    for num in arr2:
        d[num]+=1

    # Check if num not in dictionary
    for num in arr1:
        if d[num]==0:
            return num

        # Otherwise, subtract a count
        else: d[num]-=1
